### PR TITLE
SOC-1171 Purge Attribute Service URLs after update

### PR DESCRIPTION
**Pull request by jsutterfield**
Note: I closed the original PR (https://github.com/Wikia/pandora/pull/433) since unrelated code snuck in there. This is what that original PR should have looked like.

GETs to the attribute service are being cached by Fastly in varnish. This means that after a get at /user/1/attr/fbPage, an update of using a PUT or DELETE /user/1/attr/fbPage, and finally another GET of /user/1/attr/fbPage, the last GET can return stale data. After an update we want to purge those URLs from Fastly. This PR implements that.

Ticket: https://wikia-inc.atlassian.net/browse/SOC-1171

### CODE REVIEW

File: service/user-attribute/src/main/java/com/wikia/services/userattribute/guicemodule/UserAttributeModuleFactory.java [hunk 1]
```diff
@@ -9,9 +9,8 @@
 public class UserAttributeModuleFactory {
```
Comment (WikiaTech) [inline service/user-attribute/src/main/java/com/wikia/services/userattribute/guicemodule/UserAttributeModuleFactory.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1118)

File: service/user-attribute/src/main/java/com/wikia/services/userattribute/configuration/UserAttributeConfiguration.java [hunk 2]
```diff
@@ -30,4 +31,16 @@
   public JooqConfigs getDatabases() {
     return databases;
   }
+
+  @Valid
+  @NotNull
+  @JsonProperty
+  private HttpClientConfiguration httpClientConfiguration;
```
Comment (WikiaTech) [inline service/user-attribute/src/main/java/com/wikia/services/userattribute/configuration/UserAttributeConfiguration.java]: MINOR Move this variable to comply with Java Code Conventions. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1213)

File: service/user-attribute/src/main/java/com/wikia/services/userattribute/configuration/UserAttributeConfiguration.java [hunk 3]
```diff
@@ -30,4 +31,16 @@
...
+
+  @JsonProperty
+  private String serviceUrl;
```
Comment (WikiaTech) [inline service/user-attribute/src/main/java/com/wikia/services/userattribute/configuration/UserAttributeConfiguration.java]: MINOR Move this variable to comply with Java Code Conventions. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1213)

File: service/user-attribute/src/main/java/com/wikia/services/userattribute/utils/Purger.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+                @Named("service-url") String serviceUrl) {
+    this.httpClient = httpClient;
+    this.serviceUrl = serviceUrl;
+  }
+
+  public void purge(int userId, String paramName) {
+    for (String url : getUrlsToPurge(userId, paramName)) {
+      try {
+        purgeUrl(url);
+      } catch (IOException ex) {
```
Comment (WikiaTech) [inline service/user-attribute/src/main/java/com/wikia/services/userattribute/utils/Purger.java]: MAJOR Either log or rethrow this exception. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 3 major
* MINOR 2 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Define and throw a dedicated exception instead of using a generic one.](https://github.com/Wikia/pandora/blob/045d37c85abff3bd51cad87184504194e051d245/service/user-attribute/src/main/java/com/wikia/services/userattribute/guicemodule/UserAttributeModuleFactory.java#L31) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00112)

File: service/user-attribute/src/main/java/com/wikia/services/userattribute/guicemodule/UserAttributeModuleFactory.java [hunk 5]
```diff
@@ -9,25 +9,22 @@
...
 
   public static final String ENVIRONMENT = "ENVIRONMENT";
-  public static final String STAGE_PROD = "production";
-  public static final String STAGE_TESTING = "testing";
-  public static final String STAGE_INTEGRATION = "integration";
+  public static final String STAGE_PROD = "prod";
+  public static final String STAGE_DEV = "dev";
   private static final Logger LOGGER = LoggerFactory.getLogger(UserAttributeModuleFactory.class);
 
+  private UserAttributeModuleFactory() {}
```
Comment (WikiaTech) [inline service/user-attribute/src/main/java/com/wikia/services/userattribute/guicemodule/UserAttributeModuleFactory.java]: MAJOR Move this closing curly brace to the next line. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ARightCurlyBraceStartLineCheck)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Define and throw a dedicated exception instead of using a generic one.](https://github.com/Wikia/pandora/blob/3f9f37edc949670b1899b328b0d21e71f8be53a9/service/user-attribute/src/main/java/com/wikia/services/userattribute/guicemodule/UserAttributeModuleFactory.java#L33) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00112)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Define and throw a dedicated exception instead of using a generic one.](https://github.com/Wikia/pandora/blob/cf597e44312ce66c673d1b49dd0cb748142a9fce/service/user-attribute/src/main/java/com/wikia/services/userattribute/guicemodule/UserAttributeModuleFactory.java#L35) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00112)

File: service/user-attribute/src/main/java/com/wikia/services/userattribute/UserAttributeApplication.java [hunk 6]
```diff
@@ -42,7 +42,7 @@ public static void main(String[] args) throws Exception {
 
   @Override
   public Module getApplicationModule() {
-    return UserAttributeModuleFactory.createUserAttributeModule();
+    return new UserAttributeModule();
```
Comment (jsutterfield) [inline service/user-attribute/src/main/java/com/wikia/services/userattribute/UserAttributeApplication.java]: This is unrelated to the core purpose of this ticket, I'm just cleaning up the code a bit. We have a factor but it returns the same module no matter what. We can just return `UserAttributeModule` until we have a use case for needing that factory.

File: service/user-attribute/src/main/java/com/wikia/services/userattribute/utils/Purger.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+                @Named("service-url") String serviceUrl) {
+    this.httpClient = httpClient;
+    this.serviceUrl = serviceUrl;
+  }
+
+  public void purge(int userId, String paramName) {
+    for (String url : getUrlsToPurge(userId, paramName)) {
+      try {
+        purgeUrl(url);
+      } catch (IOException e) {
```
Comment (jsutterfield) [inline service/user-attribute/src/main/java/com/wikia/services/userattribute/utils/Purger.java]: There were a bunch of places below `purgeUrl()` which could throw an IOExcpetion (L58, L66, L78, and L88). These exceptions are unrelated to Fastly returning a bad response, it means something went wrong at a lower level. So I decided to catch them all here and log the error.

File: service/user-attribute/src/main/java/com/wikia/services/userattribute/utils/Purger.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+      }
+    }
+  }
+
+  private boolean responseIsBad(CloseableHttpResponse response) {
+    int statusCode = response.getStatusLine().getStatusCode();
+    return statusCode != Response.Status.OK.getStatusCode();
+  }
+
+  private void logBadResponse(String url, CloseableHttpResponse response) throws IOException {
```
Comment (rafalkalinski) [inline service/user-attribute/src/main/java/com/wikia/services/userattribute/utils/Purger.java]: And what about
> Fair point, I'll update in the new PR :)
? :)

File: service/user-attribute/src/main/java/com/wikia/services/userattribute/utils/Purger.java (hunk 7 above)
Comment (rafalkalinski) [inline service/user-attribute/src/main/java/com/wikia/services/userattribute/utils/Purger.java]: Would it be worth to write in the logger the information about where exactly from the exception was thrown?

**Comment by rafalkalinski** [comment]
Except the comments: LGTM

File: service/user-attribute/src/main/java/com/wikia/services/userattribute/utils/Purger.java (hunk 7 above)
Comment (jsutterfield) [inline service/user-attribute/src/main/java/com/wikia/services/userattribute/utils/Purger.java]: Yep, it's there already. By passing the exception into the logger after the message it formats a traceback to show you where it was thrown.

File: service/user-attribute/src/main/java/com/wikia/services/userattribute/utils/Purger.java (hunk 8 above)
Comment (jsutterfield) [inline service/user-attribute/src/main/java/com/wikia/services/userattribute/utils/Purger.java]: That's part of this comment: https://github.com/Wikia/pandora/pull/437/files#r38604282, I'm catching the exception higher up.

The logging that this method does is if we receive a bad response from Fastly, ie, that we weren't able to purge the URL. We assume that we should be able to read the response, but if something goes wrong with the connection, it's not a "Fastly couldn't purge the URL issue", it's a lower level issue. We could try and catch the IOException here, but since there's a number of different places it could be thrown during a call to `purgeUrl()`, it seemed easiest to just catch them all there to consolidate the logging.

**Comment by mklucsarits** [comment]
LGTM
